File: src/test/java/com/jinloes/prpilot/services/stream/StreamDtoTest.java
package com.jinloes.prpilot.services.stream;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

class StreamDtoTest {

    @Nested
    class StreamEventTest {

        @Test
        void type_returnsValue() {
            StreamEvent event = new StreamEvent();
            event.setType("assistant");
            assertThat(event.type()).isEqualTo("assistant");
        }

        @Test
        void message_presentWhenSet() {
            StreamEvent event = new StreamEvent();
            EventMessage msg = new EventMessage();
            event.setMessage(msg);
            assertThat(event.message()).contains(msg);
        }

        @Test
        void message_emptyWhenNull() {
            StreamEvent event = new StreamEvent();
            assertThat(event.message()).isEmpty();
        }

        @Test
        void result_presentWhenSet() {
            StreamEvent event = new StreamEvent();
            event.setResult("the result");
            assertThat(event.result()).contains("the result");
        }

        @Test
        void result_emptyWhenNull() {
            StreamEvent event = new StreamEvent();
            assertThat(event.result()).isEmpty();
        }
    }

    @Nested
    class EventMessageTest {

        @Test
        void content_presentWhenSet() {
            EventMessage msg = new EventMessage();
            ContentBlock block = new ContentBlock();
            msg.setContent(List.of(block));
            assertThat(msg.content()).contains(List.of(block));
        }

        @Test
        void content_emptyWhenNull() {
            EventMessage msg = new EventMessage();
            assertThat(msg.content()).isEmpty();
        }
    }

    @Nested
    class ContentBlockTest {

        @Test
        void type_returnsValue() {
            ContentBlock block = new ContentBlock();
            block.setType("tool_use");
            assertThat(block.type()).isEqualTo("tool_use");
        }

        @Test
        void name_presentWhenSet() {
            ContentBlock block = new ContentBlock();
            block.setName("mcp__github__get_file");
            assertThat(block.name()).contains("mcp__github__get_file");
        }

        @Test
        void name_emptyWhenNull() {
            ContentBlock block = new ContentBlock();
            assertThat(block.name()).isEmpty();
        }

        @Test
        void input_presentWhenSet() {
            ContentBlock block = new ContentBlock();
            Map<String, Object> input = Map.of("owner", "alice");
            block.setInput(input);
            assertThat(block.input()).contains(input);
        }

        @Test
        void input_emptyWhenNull() {
            ContentBlock block = new ContentBlock();
            assertThat(block.input()).isEmpty();
        }
    }
}

File: core/src/jvmTest/kotlin/com/jinloes/prpilot/services/stream/StreamDtoTest.kt
package com.jinloes.prpilot.services.stream

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

class StreamDtoTest : FunSpec({

    // ── StreamEvent ──────────────────────────────────────────────────────

    context("StreamEvent") {

        test("type returns set value") {
            StreamEvent(type = "assistant").type shouldBe "assistant"
        }

        test("type returns null when not set") {
            StreamEvent().type.shouldBeNull()
        }

        test("subtype returns set value") {
            StreamEvent(subtype = "success").subtype shouldBe "success"
        }

        test("subtype returns null when not set") {
            StreamEvent().subtype.shouldBeNull()
        }

        test("isError defaults to false") {
            StreamEvent().isError.shouldBeFalse()
        }

        test("isError returns true when set") {
            StreamEvent(isError = true).isError.shouldBeTrue()
        }

        test("message is non-null when set") {
            val msg = EventMessage()
            StreamEvent(message = msg).message.shouldNotBeNull()
        }

        test("message is null when not set") {
            StreamEvent().message.shouldBeNull()
        }

        test("result is non-null when set") {
            StreamEvent(result = "the result").result shouldBe "the result"
        }

        test("result is null when not set") {
            StreamEvent().result.shouldBeNull()
        }

        test("sessionId returns value when set") {
            StreamEvent(sessionId = "abc-123").sessionId shouldBe "abc-123"
        }

        test("sessionId returns null when not set") {
            StreamEvent().sessionId.shouldBeNull()
        }
    }

    // ── EventMessage ─────────────────────────────────────────────────────

    context("EventMessage") {

        test("content is non-null when set") {
            val block = ContentBlock()
            EventMessage(content = listOf(block)).content shouldBe listOf(block)
        }

        test("content is null when not set") {
            EventMessage().content.shouldBeNull()
        }
    }

    // ── ContentBlock ─────────────────────────────────────────────────────

    context("ContentBlock") {

        test("type returns set value") {
            ContentBlock(type = "tool_use").type shouldBe "tool_use"
        }

        test("type returns null when not set") {
            ContentBlock().type.shouldBeNull()
        }

        test("name is non-null when set") {
            ContentBlock(name = "mcp__github__get_file").name shouldBe "mcp__github__get_file"
        }

        test("name is null when not set") {
            ContentBlock().name.shouldBeNull()
        }

        test("input is non-null when set") {
            val input = mapOf("owner" to "alice" as Any)
            ContentBlock(input = input).input shouldBe input
        }

        test("input is null when not set") {
            ContentBlock().input.shouldBeNull()
        }

        test("text is non-null when set") {
            ContentBlock(text = "hello").text shouldBe "hello"
        }

        test("text is null when not set") {
            ContentBlock().text.shouldBeNull()
        }

        test("thinking is non-null when set") {
            ContentBlock(thinking = "deep thought").thinking shouldBe "deep thought"
        }

        test("thinking is null when not set") {
            ContentBlock().thinking.shouldBeNull()
        }
    }

    // ── AnySerializer ────────────────────────────────────────────────────

    context("AnySerializer.fromElement") {

        test("JsonNull deserializes to 'null' string") {
            AnySerializer.fromElement(JsonNull) shouldBe "null"
        }

        test("string primitive deserializes to String") {
            val result = AnySerializer.fromElement(JsonPrimitive("hello"))
            result shouldBe "hello"
            (result is String).shouldBeTrue()
        }

        test("boolean primitive true deserializes to Boolean") {
            val result = AnySerializer.fromElement(JsonPrimitive(true))
            result shouldBe true
            (result is Boolean).shouldBeTrue()
        }

        test("boolean primitive false deserializes to Boolean") {
            val result = AnySerializer.fromElement(JsonPrimitive(false))
            result shouldBe false
            (result is Boolean).shouldBeTrue()
        }

        test("long primitive deserializes to Long") {
            val result = AnySerializer.fromElement(JsonPrimitive(42L))
            (result is Long).shouldBeTrue()
            result shouldBe 42L
        }

        test("double primitive deserializes to Double") {
            val result = AnySerializer.fromElement(JsonPrimitive(3.14))
            (result is Double).shouldBeTrue()
        }

        test("JsonObject deserializes to Map") {
            val obj: JsonObject = buildJsonObject { put("key", JsonPrimitive("val")) }
            val result = AnySerializer.fromElement(obj)
            (result is Map<*, *>).shouldBeTrue()
            @Suppress("UNCHECKED_CAST")
            (result as Map<String, Any>)["key"] shouldBe "val"
        }
    }
})
